cpp/dolfinx/graph/AdjacencyList.h

```diff
@@ -32,83 +50,49 @@ class AdjacencyList
   /// @param [in] n Number of nodes
   explicit AdjacencyList(const std::int32_t n) : _array(n), _offsets(n + 1)
   {
-    std::iota(_array.data(), _array.data() + n, 0);
-    std::iota(_offsets.data(), _offsets.data() + n + 1, 0);
+    std::iota(_array.begin(), _array.end(), 0);
+    std::iota(_offsets.begin(), _offsets.end(), 0);
   }
 
-  /// Construct adjacency list from arrays of data (Eigen data types)
+  /// Construct adjacency list from arrays of data
   /// @param [in] data Adjacency array
   /// @param [in] offsets The index to the adjacency list in the data
-  ///   array for node i
-  template <typename U, typename V,
-            std::enable_if_t<std::is_base_of<Eigen::EigenBase<std::decay_t<V>>,
-                                             std::decay_t<V>>::value,
-                             int> = 0>
+  /// array for node i
+  template <typename U, typename V>
   AdjacencyList(U&& data, V&& offsets)
       : _array(std::forward<U>(data)), _offsets(std::forward<V>(offsets))
   {
-    // Do nothing
-  }
-
-  /// Construct adjacency list from arrays of data  (non-Eigen data
-  /// types)
-  /// @param [in] data Adjacency array
-  /// @param [in] offsets The index to the adjacency list in the data
-  ///   array for node i
-  template <typename U, typename V,
-            std::enable_if_t<!std::is_base_of<Eigen::EigenBase<std::decay_t<V>>,
-                                              std::decay_t<V>>::value,
-                             int> = 0>
-  AdjacencyList(U&& data, V&& offsets)
-      : _array(data.size()), _offsets(offsets.size())
-  {
-    assert(offsets.back() == (std::int32_t)data.size());
-    std::copy(data.begin(), data.end(), _array.data());
-    std::copy(offsets.begin(), offsets.end(), _offsets.data());
+    assert(_offsets.back() == (std::int32_t)_array.size());
   }
 
+  /// @todo To be removed
   /// Construct adjacency list for a problem with a fixed number of
   /// links (edges) for each node
   /// @param [in] matrix Two-dimensional array of adjacency data where
-  ///   matrix(i, j) is the jth neighbor of the ith node
+  /// matrix(i, j) is the jth neighbor of the ith node
   explicit AdjacencyList(
       const Eigen::Ref<const Eigen::Array<T, Eigen::Dynamic, Eigen::Dynamic,
                                           Eigen::RowMajor>>& matrix)
-      : _array(matrix.rows() * matrix.cols()), _offsets(matrix.rows() + 1)
   {
-    const std::int32_t num_links = matrix.cols();
-    for (Eigen::Index e = 0; e < _offsets.rows(); e++)
-      _offsets[e] = e * num_links;
-
-    // NOTE: Do not directly copy data from matrix because it may be a
-    // view into a larger array
-    for (Eigen::Index i = 0; i < matrix.rows(); ++i)
-      _array.segment(_offsets(i), num_links) = matrix.row(i);
+    std::tie(_array, _offsets) = create_adjacency_data(matrix);
```

Comment:
Is `std::tie` preferred to structured binding here?

Reply:
I'm not sure what the options are. Structured binding followed by a move?

This constructor will get removed at some point in favour of a helper function - keeps the constructors simple and computation light.

Reply:
Ok, I looked for some options, and `std::tie` seems to be neater for variables already declared.